tests/unit/cluster/auto-failover-on-shutdown.tcl

```diff
@@ -0,0 +1,72 @@
+proc shutdown_on_how {srv_id how} {
+    if {$how == "shutdown"} {
+        catch {R $srv_id shutdown nosave}
+    } elseif {$how == "sigterm"} {
+        exec kill -SIGTERM [s -$srv_id process_id]
+    }
+}
+
+# We will start a cluster with 3 primary nodes and 4 replicas, the primary 1 will have 2 replicas.
+# We will pause the replica 1, and then shutdown the primary 1, and making replica 2 to become
+# the new primary.
+proc test_main {how} {
+    test "auto-failover-on-shutdown will always pick a best replica and send CLUSTER FAILOVER - $how" {
+        set primary [srv 0 client]
+        set replica1 [srv -3 client]
+        set replica1_pid [s -3 process_id]
+        set replica2 [srv -6 client]
+        set replica2_ip [srv -6 host]
+        set replica2_port [srv -6 port]
+
+        # Pause a replica so it has no chance to catch up with the offset.
+        pause_process $replica1_pid
+
+        # Primary write some data to increase the offset.
+        for {set i 0} {$i < 10} {incr i} {
+            $primary incr key_991803
+        }
+
+        # Wait the replica2 catch up with the offset
+        wait_for_ofs_sync $primary $replica2
```

Comment:
I wish the test can pass even without this wait. If the primary has many writes and receives SIGTERM, we want failover without data loss.

The primary does pause write in prepare shutdown and it waits for replicas offset before finish shutdown, but if we failover immediately, the replica can miss some writes before the failover?

I think we need to postpone failover and do it when the offset is synced, some time between prepare shutdown and finish shutdown, or maybe in finish shutdown as in your first implementation, but don't wait for all replicas when this feature is enabled?

Every implementation has pros and cons?

Reply:
The test does not enable shutdown-timeout (it is 0 in the test), so the primary won't pause write and wait for the replica in this shutdown-timeout part.

Maybe we can, if the offset is equal to primary_repl_offset, we use a FORCE failover, if it is not, we use a plain failover? Noted that we will call flushReplicasOutputBuffers before doing this failover thing and exit thing, so i guess the replicas may still have a chance to catch up with the offset.

i am thinking it we enable shutdown-timeout, we have a good chance that can find a best replica, if we are not, i don't know if we should wait?

i will try to stable the test and also add a test with shutdown-timeout enabled.

Reply:
> The test does not enable shutdown-timeout (it is 0 in the test), so the primary won't pause write and wait for the replica in this shutdown-timeout part.

I think this feature makes more sense with shutdown-timeout. If you care about data loss, then you don't want to disable shutdown-timeout.

> Maybe we can, if the offset is equal to primary_repl_offset, we use a FORCE failover, if it is not, we use a plain failover?

Interesting idea, but more complex? We get more different cases.

> Noted that we will call flushReplicasOutputBuffers before doing this failover thing and exit thing, so i guess the replicas may still have a chance to catch up with the offset.

yes, but now in your code we send CLUSTER FAILOVER only if the replica is already in sync. Otherwise, we will not send it?

```C
            /* This is done only when the replica offset is caught up, to avoid data loss */
            if (replica->repl_state == REPLICA_STATE_ONLINE && replica->repl_ack_off == server.primary_repl_offset) {
                best_replica = replica;
                break;
            }
```

> i am thinking it we enable shutdown-timeout, we have a good chance that can find a best replica, if we are not, i don't know if we should wait?

I'm not sure what you mean. If we enable shutdown-timeout, we can wait, right? If shutdown-timeout is 0, we don't wait. But let's design it for shutdown-timeout > 0, because it makes more sense?

I don't see the problem with waiting. We have the `isReadyToShutdown` that we check in cron while we're waiting. When we have one replica up to date, we can send the CLUSTER FAILOVER FORCE to it. What's the problem with this idea? We need to set a flag like `server.shutdown_failover_sent = 1` to make sure we don't do it twice? Sure...

Reply:
Yes, i think we can make shutdown-failover only take effect when shutdown-timeout is enabled. 

I did try shutdown_failover_sent before, what should we do if the old primary received a UPDATE, should we turn myself into a replica of the new primary? or should we ignore it and just finish the shutdown. Since the waitting time is enough, currently the old primary will become a replica if it see an UPDATE, and then continue the finishShutdown thing. (i am not happy with this become replica thing... sure we can somehow check the shutdown_failover_sent thing in cluster.c to skip some replica thing, but it is odd)

Reply:
> I did try shutdown_failover_sent before, what should we do if the old primary received a UPDATE, should we turn myself into a replica of the new primary?

If we receive an update, we can go immediately to finishShutdown? We don't need to wait for other replicas after that, right?

If we turn into a replica or not is not important as long as there is no data loss. :laughing:

Reply:
yes, thinking it again, it seem that becoming a replica is not a big deal, since it can always do a psync so it is not a big cost. I will handle it later.